### ISSUE TITLE
[tracking] Facade

### ISSUE THREAD

**Issue by jsdw** [paritytech/polkadot-sdk#6441]
This is a tracking issue which collects andy issues that are releated to the Facade project.

```[tasklist]
### Tasks
- [ ] #6396
- [ ] #4722 
```

**Comment by xlc** [comment]
what is this facade project?

**Comment by xlc** [comment]
should XCQ be part of this? https://github.com/polkadot-fellows/RFCs/pull/126

**Comment by jsdw** [comment]
> Fascade itself should be written in Rust. Rust itself can be integrated almost anywhere. Like you said into typescript via WASM or directly into C, C++, python or whatever. The fascade Rust library should do all the heavy lifting, aka the decoding and ensuring there is a stable api etc.

I'm mostly onboard with that! Given that most people using it will write TS apps, I wonder if it would make sense though to write a separate TS and Rust impl of the Facade Client library. That way, the TS version will be very idiomatic, have a small bundle size and be written in PAPI, which already has the relevant primitives for decoding etc. It'll make sense for most dApps and such.

The Rust version would then use Subxt and be the thing we bind to in other languages (and we could eventually maintain such bindings for a few langs).

Both would likely rely on the Rust stuff for historic block decoding, but this would only need to be pulled in in TS land when needed to avoid the bloat of the WASM blob.

**Comment by bkchr** [comment]
You could probably still provide some idiomatic TS library that uses the interfaces of the Rust library in a more typescript way. Same is for example done in Rust when you wrap a C library, you also expose Rust interfaces while under the hood using the C functions. But still you only write glue code in Rust and the actual implementation is in C. Same principle could be applied here.

**Comment by xlc** [comment]
XCQ extension already handles the API definition part and it provides metadata so client can generate whatever code it needs from the metadata

**Comment by jsdw** [comment]
> should XCQ be part of this? https://github.com/polkadot-fellows/RFCs/pull/126

My understanding at the moment is that this and XCQ are complementory: 

- This facade work is like the foundation and "basic" APIs for doing some common things. It will lead to more Pallet View Functions being created and then, from these, additional Runtime APIs which use these to gather & return commonly wanted information. These Runtime APIs can be specced and implemented on many runtimes, so that a simple higher level client can call them across runtimes and aggregate the result. This work trades flexibility for simplicity; doing common things across supporting runtimes will be easy via the new APIs, but doing complex/custom things is still hard (still need to make a bunch of calls to gather data and and calculate things locally if the existing APIs don't give what you want).

- The XCQ proposal, IIUC, can then take advantage of these Pallet View Functions / Runtime APIs that we add here and allows people to build arbitrary PolkaVM programs to obtain more custom/complex data that isn't otherwise expressed by the APIs we add here in the Facade project. In other words, it covers the rest of the cases that the facade doesn't in exchange for the increased complexity of writing PolkaVM programs to achieve this.

**Comment by xlc** [comment]
yes I just want to ensure this project is designed and built in mind of XCQ to ensure they can nicely integrate together and avoid duplicate work
the client code generation for example, is not something we plan for XCQ but sounds like could benefit XCQ users if it supports XCQ

**Comment by jsdw** [comment]
> yes I just want to ensure this project is designed and built in mind of XCQ to ensure they can nicely integrate together and avoid duplicate work

Yup that makes sense! 

I think it'll be a net win overall to have both the "easy" APIs that users can use to obtain basic data across chains, plus a more advanced way to obtain and aggregate more custom/complex queries via XCQ programs. I can also imagine that when XCQ is ready we might start using it to aggregate some of this data together in place of traditional Runtime APIs where it makes sense!

Ultimately this work here shouldn't duplicate any XCQ related work. Instead. it'll provide the Pallet View Functions and new Runtime APIs which can then be fed into XCQ programs.
